models/User.go

```diff
@@ -1414,69 +1415,151 @@ func PerformMortgageTransaction(userId, stockId uint32, stockQuantity int64) (*T
 
 	l.Infof("Taking current price of stock as %d", mortgagePrice)
 
-	var rate int32
-
-	if stockQuantity >= 0 {
-		rate = MORTGAGE_RETRIEVE_RATE
-		l.Debugf("stockQuantity positive. Retrieving stocks from mortgage @ %d%%", rate)
-
+	if stockQuantity >= 0 /* Retrieve stocks action*/ {
 		db := getDB()
 
+		l.Debugf("Retrieving stocks in action")
+
 		stockCount := struct{ Sc int64 }{0}
-		sql := "Select sum(StockQuantity) as sc from Transactions where UserId=? and StockId=? and Type=?"
-		err = db.Raw(sql, user.Id, stockId, MortgageTransaction.String()).Scan(&stockCount).Error
+		sql := "Select sum(stocksInBank) as sc from MortgageDetails where UserId=? and StockId=?"
+		err = db.Raw(sql, user.Id, stockId).Scan(&stockCount).Error
 		if err != nil {
+			l.Infof("PerformMortgageTransaction failed for userId = %d, stockId = %d amount = %d while retrieving total stocks", userId, stockId, stockQuantity)
 			l.Error(err)
 			return nil, err
 		}
 
-		// Sc will be negative if stocks are there in mortgage.
-		// Change sign to mean number of stocks in mortgage
-		stockCount.Sc *= -1
-
-		l.Infof("%d stocks mortgaged currently", stockCount.Sc)
+		l.Debugf("%d stocks mortgaged currently", stockCount.Sc)
 
 		if stockQuantity > stockCount.Sc {
 			l.Errorf("Insufficient stocks in mortgage. Have %d, want %d", stockCount.Sc, stockQuantity)
 			return nil, NotEnoughStocksError{}
 		}
 
 	} else {
-		MortgagePutLimitRWMutex.RLock()
-		lim := int64(MortgagePutLimit)
-		MortgagePutLimitRWMutex.RUnlock()
-		if user.Total > lim {
-			return nil, WayTooMuchCashError{}
-		}
+		l.Debugf("Mortgaging stocks in action")
+
+		db := getDB()
 
-		rate = MORTGAGE_DEPOSIT_RATE
-		l.Debugf("stockQuantity negative. Depositing stocks to mortgage @ %d%%", rate)
+		/* We don't want to include Mortgage type transaction because mortgage transactions are stored in Transactions
```

Comment:
I don't think it's a good idea to make this assumption. Throughout the code we have made one important invariant - `Select sum(StockQuantity) From Transactions where UserId=1 and StockId=2` should **always** give stocks that he currently owns, regardless of the type of the transaction (similar case for `Total` field in the transaction).

While the assumption you're making is consistent with the above "rule", I still feel it's not aligned with it exactly (which is why you're making two queries below to get unmortgaged stocks).